File: backend/billing/payment_engine.py
"""
UNNATI Payment Engine: Direct Service-Provider Payment & Cooperative Economics
-------------------------------------------------------------------------------
Core principle: UNNATI is NOT a financial intermediary or escrow holder.
The payment relationship is directly between customer and worker/collective.
"""

from decimal import Decimal, ROUND_HALF_UP
from datetime import datetime, timezone as dt_timezone
from typing import Dict, Any, Optional, Tuple

COOPERATIVE_RATE_DEFAULT = Decimal('0.065') # 6.5% transparent cooperative reserve
PLATFORM_FEE_DIRECT = Decimal('0.00')       # Exactly 0.00 - zero escrow, zero middleman rake

# Canonical Lifecycle States
VALID_PAYMENT_STATES = {
    'PAYMENT_PENDING',
    'PAYMENT_INITIATED',
    'PAYMENT_CONFIRMED',
    'PAYMENT_COMPLETED',
    'PAYMENT_FAILED',
    'PAYMENT_CANCELLED',
    'REFUND_PENDING',
    'REFUNDED',
}

# Permitted state transitions
ALLOWED_PAYMENT_TRANSITIONS = {
    'PAYMENT_PENDING': {'PAYMENT_INITIATED', 'PAYMENT_CANCELLED', 'PAYMENT_FAILED'},
    'PAYMENT_INITIATED': {'PAYMENT_CONFIRMED', 'PAYMENT_COMPLETED', 'PAYMENT_FAILED', 'PAYMENT_CANCELLED'},
    'PAYMENT_CONFIRMED': {'PAYMENT_COMPLETED', 'REFUND_PENDING'},
    'PAYMENT_COMPLETED': {'REFUND_PENDING'},
    'PAYMENT_FAILED': {'PAYMENT_INITIATED', 'PAYMENT_CANCELLED'},
    'REFUND_PENDING': {'REFUNDED', 'PAYMENT_COMPLETED'},
    'PAYMENT_CANCELLED': set(),
    'REFUNDED': set(),
}


def calculate_direct_payment_breakdown(
    amount: Decimal,
    worker_count: int = 1,
    cooperative_rate: Decimal = COOPERATIVE_RATE_DEFAULT
) -> Dict[str, Any]:
    """
    Computes transparent server-side breakdown for direct worker payment.
    Ensures 0 platform middleman fee and verifies no platform escrow.
    """
    amount = Decimal(str(amount)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
    worker_count = max(1, int(worker_count))

    # Cooperative contribution (retained by cooperative entity for member patronage dividend)
    coop_allocation = (amount * cooperative_rate).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
    
    # Direct Worker Payout is the total payable less the cooperative dividend reserve
    worker_direct_payout = (amount - coop_allocation).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
    
    # Equal split per worker for collective crew bookings
    per_worker_share = (worker_direct_payout / Decimal(worker_count)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
    
    # Total payout reconciled with rounding adjustment on first worker
    rounding_diff = worker_direct_payout - (per_worker_share * Decimal(worker_count))
    first_worker_share = per_worker_share + rounding_diff

    breakdown = {
        'total_customer_paid': amount,
        'worker_direct_payout': worker_direct_payout,
        'cooperative_allocation': coop_allocation,
        'cooperative_rate_percentage': float(cooperative_rate * 100),
        'platform_fee': PLATFORM_FEE_DIRECT,
        'platform_escrow_balance': Decimal('0.00'),
        'platform_holds_escrow': False,
        'worker_count': worker_count,
        'per_worker_share': per_worker_share,
        'first_worker_share': first_worker_share,
        'is_collective': worker_count > 1,
    }
    
    ensure_no_platform_escrow(breakdown)
    return breakdown


def ensure_no_platform_escrow(breakdown: Dict[str, Any]) -> None:
    """
    Security check: raises ValueError if platform fee is non-zero or platform holds funds.
    """
    if breakdown.get('platform_fee') != Decimal('0.00'):
        raise ValueError("UNNATI policy violation: Platform fee must be 0.00 in direct model.")
    if breakdown.get('platform_escrow_balance') != Decimal('0.00'):
        raise ValueError("UNNATI policy violation: Platform escrow balance must be 0.00.")
    if breakdown.get('platform_holds_escrow') is not False:
        raise ValueError("UNNATI policy violation: Platform must never hold customer escrow.")


def validate_payment_state_transition(
    current_state: str,
    next_state: str,
    user_role: str,
    is_worker_confirmed: bool = False
) -> Tuple[bool, Optional[str]]:
    """
    Enforces server-side authorization and transition rules for payment lifecycle.
    """
    if current_state not in VALID_PAYMENT_STATES:
        return False, f"Invalid current payment state: {current_state}"
    if next_state not in VALID_PAYMENT_STATES:
        return False, f"Invalid target payment state: {next_state}"
        
    allowed_targets = ALLOWED_PAYMENT_TRANSITIONS.get(current_state, set())
    if next_state not in allowed_targets:
        return False, f"Transition from '{current_state}' to '{next_state}' is not permitted."

    # Role-based authorization rules
    if next_state in {'PAYMENT_INITIATED', 'PAYMENT_CANCELLED'} and current_state == 'PAYMENT_PENDING':
        if user_role not in {'customer', 'admin'}:
            return False, "Only customer or admin can initiate or cancel pending payment."

    if next_state in {'PAYMENT_CONFIRMED', 'PAYMENT_COMPLETED'}:
        # Must be confirmed by worker or provider adapter / admin
        if user_role not in {'worker', 'admin'} and not is_worker_confirmed:
            return False, "Only service provider (worker), provider adapter, or admin can confirm direct payment receipt."

    if next_state == 'REFUND_PENDING':
        if user_role not in {'customer', 'worker', 'admin'}:
            return False, "Unauthorized refund request."

    if next_state == 'REFUNDED':
        if user_role not in {'worker', 'admin'}:
            return False, "Refund execution must be verified by the direct recipient worker or admin."

    return True, None


def calculate_cancellation_refund(
    total_amount: Decimal,
    booking_status: str,
    created_at: datetime,
    scheduled_time: Optional[datetime] = None,
    now: Optional[datetime] = None
) -> Dict[str, Any]:
    """
    Integrates existing booking cancellation safeguards into direct payment records.
    - Free cancellation in 5 min grace period or >2 hours ahead for scheduled.
    - ₹100 compensation if worker on the way.
    - ₹200 compensation if worker arrived at job site.
    - Remainder is refunded directly to customer.
    """
    now = now or datetime.now(dt_timezone.utc)
    total_amount = Decimal(str(total_amount))
    
    # 5-minute grace period
    grace_period_seconds = 300
    seconds_since_creation = (now - created_at).total_seconds()
    is_within_grace = seconds_since_creation <= grace_period_seconds

    # Scheduled booking >2h buffer
    is_advance_scheduled_free = False
    if scheduled_time:
        seconds_until_scheduled = (scheduled_time - now).total_seconds()
        if seconds_until_scheduled >= 7200:
            is_advance_scheduled_free = True

    status_normalized = (booking_status or '').lower()
    
    if is_within_grace or is_advance_scheduled_free or status_normalized in ['requested', 'matching', 'searching', 'accepted', 'scheduled']:
        compensation = Decimal('0.00')
        reason = "Free cancellation within grace window"
    elif status_normalized in ['worker_arriving', 'on_the_way']:
        compensation = min(Decimal('100.00'), total_amount)
        reason = "Progressive compensation: worker dispatched / on the way"
    elif status_normalized in ['arrived', 'in_progress', 'repair_started']:
        compensation = min(Decimal('200.00'), total_amount)
        reason = "Progressive compensation: worker arrived on site"
    else:
        compensation = Decimal('0.00')
        reason = "Standard refund"

    refund_amount = max(Decimal('0.00'), total_amount - compensation)
    
    return {
        'total_amount': total_amount,
        'cancellation_compensation': compensation,
        'refund_amount': refund_amount,
        'reason': reason,
        'platform_retained': Decimal('0.00'),
    }


def generate_upi_intent_uri(
    vpa: str,
    payee_name: str,
    amount: Decimal,
    booking_id: int,
    transaction_ref: Optional[str] = None
) -> str:
    """
    Generates standard Indian NPCI UPI intent URI for direct customer-to-worker payment.
    Enables instant checkout via Google Pay, PhonePe, Paytm, BHIM, etc.
    """
    clean_name = payee_name.replace(' ', '%20')
    ref = transaction_ref or f"UNN-{booking_id}"
    amt_str = f"{amount:.2f}"
    return f"upi://pay?pa={vpa}&pn={clean_name}&am={amt_str}&cu=INR&tn=UNNATI-Booking-{booking_id}&tr={ref}"

File: frontend/src/lib/paymentEngine.ts
/**
 * UNNATI Payment Engine: Direct Service-Provider Payment & Cooperative Economics
 * -------------------------------------------------------------------------------
 * Core Rule: UNNATI must NOT act as a financial middleman or hold customer money in platform escrow.
 * The payment relationship is strictly direct between customer and service provider.
 */

import {
  PaymentLifecycleStatus,
  DirectPaymentBreakdown,
  WorkerAllocationShare
} from '../types/unnati';

export const UNNATI_COOPERATIVE_RATE = 0.065; // 6.5% transparent cooperative reserve
export const PLATFORM_FEE_DIRECT = 0.0;       // Exactly 0.00 - zero escrow, zero middleman rake

export const ALLOWED_PAYMENT_TRANSITIONS: Record<PaymentLifecycleStatus, PaymentLifecycleStatus[]> = {
  PAYMENT_PENDING: ['PAYMENT_INITIATED', 'PAYMENT_CANCELLED', 'PAYMENT_FAILED'],
  PAYMENT_INITIATED: ['PAYMENT_CONFIRMED', 'PAYMENT_COMPLETED', 'PAYMENT_FAILED', 'PAYMENT_CANCELLED'],
  PAYMENT_CONFIRMED: ['PAYMENT_COMPLETED', 'REFUND_PENDING'],
  PAYMENT_COMPLETED: ['REFUND_PENDING'],
  PAYMENT_FAILED: ['PAYMENT_INITIATED', 'PAYMENT_CANCELLED'],
  REFUND_PENDING: ['REFUNDED', 'PAYMENT_COMPLETED'],
  PAYMENT_CANCELLED: [],
  REFUNDED: [],
};

/**
 * Calculates transparent client-side and server-side matching breakdown.
 * Ensures platform fee is zero and platform never holds escrow balance.
 */
export function calculateDirectPaymentBreakdown(
  amount: number,
  workerCount = 1,
  cooperativeRate = UNNATI_COOPERATIVE_RATE,
  workerName = 'Assigned Craftsman',
  workerVpa?: string,
  assignedWorkers?: Array<{ id: number; full_name: string }>
): DirectPaymentBreakdown {
  const safeAmount = Math.max(0, Math.round(amount * 100) / 100);
  const safeCount = Math.max(1, Math.floor(workerCount));

  // 6.5% transparent cooperative reserve
  const cooperativeAllocation = Math.round(safeAmount * cooperativeRate * 100) / 100;
  
  // Direct Worker Payout is total customer paid minus the cooperative reserve
  const workerDirectPayout = Math.round((safeAmount - cooperativeAllocation) * 100) / 100;

  // Split equally per worker for collective crew bookings
  const perWorkerShare = Math.round((workerDirectPayout / safeCount) * 100) / 100;
  const roundingDiff = Math.round((workerDirectPayout - perWorkerShare * safeCount) * 100) / 100;
  const firstWorkerShare = Math.round((perWorkerShare + roundingDiff) * 100) / 100;

  // Build worker shares if assigned workers are present
  const workerShares: WorkerAllocationShare[] = [];
  if (assignedWorkers && assignedWorkers.length > 0) {
    assignedWorkers.forEach((w, idx) => {
      workerShares.push({
        workerId: w.id,
        workerName: w.full_name,
        allocatedPayout: idx === 0 ? firstWorkerShare : perWorkerShare,
        cooperativeDividendShare: Math.round((cooperativeAllocation / safeCount) * 100) / 100,
        status: 'assigned',
      });
    });
  }

  const breakdown: DirectPaymentBreakdown = {
    totalCustomerPaid: safeAmount,
    workerDirectPayout,
    cooperativeAllocation,
    cooperativeRatePercentage: Math.round(cooperativeRate * 1000) / 10,
    platformFee: PLATFORM_FEE_DIRECT,
    platformEscrowBalance: 0.0,
    platformHoldsEscrow: false,
    workerCount: safeCount,
    perWorkerShare,
    firstWorkerShare,
    workerName,
    workerVpa,
    isCollective: safeCount > 1,
    workerShares: workerShares.length > 0 ? workerShares : undefined,
  };

  verifyZeroPlatformEscrow(breakdown);
  return breakdown;
}

/**
 * Validates strictly that platform holds 0 escrow and charges 0 fee.
 */
export function verifyZeroPlatformEscrow(breakdown: DirectPaymentBreakdown): void {
  if (breakdown.platformFee !== 0.0) {
    throw new Error('Security Violation: Platform fee must be 0.00 in UNNATI direct payment model.');
  }
  if (breakdown.platformEscrowBalance !== 0.0) {
    throw new Error('Security Violation: Platform escrow balance must be 0.00.');
  }
  if (breakdown.platformHoldsEscrow !== false) {
    throw new Error('Security Violation: UNNATI must never hold platform escrow.');
  }
}

/**
 * Validates payment state machine transitions with role authorizations.
 */
export function validatePaymentStateTransition(
  current: PaymentLifecycleStatus,
  next: PaymentLifecycleStatus,
  role: string,
  isWorkerConfirmed = false
): { valid: boolean; error?: string } {
  const allowed = ALLOWED_PAYMENT_TRANSITIONS[current] || [];
  if (!allowed.includes(next)) {
    return {
      valid: false,
      error: `Transition from '${current}' to '${next}' is not allowed.`,
    };
  }

  // Authorization checks
  if ((next === 'PAYMENT_INITIATED' || next === 'PAYMENT_CANCELLED') && current === 'PAYMENT_PENDING') {
    if (role !== 'customer' && role !== 'admin') {
      return { valid: false, error: 'Only customer or admin can initiate or cancel pending payment.' };
    }
  }

  if (next === 'PAYMENT_CONFIRMED' || next === 'PAYMENT_COMPLETED') {
    if (role !== 'worker' && role !== 'admin' && !isWorkerConfirmed) {
      return { valid: false, error: 'Only service provider or verified adapter can confirm direct payment receipt.' };
    }
  }

  if (next === 'REFUNDED') {
    if (role !== 'worker' && role !== 'admin') {
      return { valid: false, error: 'Refund must be confirmed by service provider or admin.' };
    }
  }

  return { valid: true };
}

/**
 * Generates an Indian NPCI UPI intent URI for direct customer-to-worker payment.
 */
export function generateUpiIntentUrl(params: {
  vpa: string;
  payeeName: string;
  amount: number;
  bookingId: number;
  ref?: string;
}): string {
  const cleanName = encodeURIComponent(params.payeeName);
  const ref = params.ref || `UNN-${params.bookingId}`;
  const amtStr = params.amount.toFixed(2);
  return `upi://pay?pa=${params.vpa}&pn=${cleanName}&am=${amtStr}&cu=INR&tn=UNNATI-Booking-${params.bookingId}&tr=${ref}`;
}

/**
 * Computes direct refund and worker compensation based on booking cancellation stage.
 */
export function calculateDirectCancellationRefund(params: {
  totalAmount: number;
  bookingStatus: string;
  createdAt: string | number | Date;
  scheduledTime?: string | number | Date | null;
  now?: number;
}): {
  refundAmount: number;
  workerCompensation: number;
  isFree: boolean;
  explanation: string;
  platformRetained: number;
} {
  const now = params.now || Date.now();
  const created = new Date(params.createdAt).getTime();
  const totalAmount = Math.max(0, params.totalAmount);

  // 5 minute grace period
  const gracePeriodMs = 5 * 60 * 1000;
  const isWithinGrace = now - created <= gracePeriodMs;

  // Scheduled booking >2h buffer
  let isAdvanceScheduledFree = false;
  if (params.scheduledTime) {
    const sched = new Date(params.scheduledTime).getTime();
    if (sched - now >= 2 * 60 * 60 * 1000) {
      isAdvanceScheduledFree = true;
    }
  }

  const status = (params.bookingStatus || '').toLowerCase();

  let workerCompensation = 0;
  let explanation = '';
  let isFree = false;

  if (
    isWithinGrace ||
    isAdvanceScheduledFree ||
    ['requested', 'matching', 'searching', 'accepted', 'scheduled'].includes(status)
  ) {
    workerCompensation = 0;
    isFree = true;
    explanation = isWithinGrace
      ? 'Free cancellation within 5-minute grace period.'
      : isAdvanceScheduledFree
      ? 'Free cancellation: Scheduled >2 hours in advance.'
      : 'Free cancellation prior to worker departure.';
  } else if (['worker_arriving', 'on_the_way'].includes(status)) {
    workerCompensation = Math.min(100, totalAmount);
    isFree = false;
    explanation = '₹100 dispatch compensation credited directly to worker for on-the-way cancellation.';
  } else if (['arrived', 'in_progress', 'repair_started'].includes(status)) {
    workerCompensation = Math.min(200, totalAmount);
    isFree = false;
    explanation = '₹200 on-site compensation credited directly to worker for cancellation after arrival.';
  } else {
    workerCompensation = 0;
    isFree = true;
    explanation = 'Standard cancellation.';
  }

  const refundAmount = Math.max(0, Math.round((totalAmount - workerCompensation) * 100) / 100);

  return {
    refundAmount,
    workerCompensation,
    isFree,
    explanation,
    platformRetained: 0, // Always 0 platform middleman fee
  };
}

/**
 * Returns mobile-first visual progress metadata for UNNATI payment lifecycle.
 */
export function getPaymentLifecycleProgress(status: PaymentLifecycleStatus): {
  percent: number;
  label: string;
  labelHi: string;
  colorClass: string;
} {
  switch (status) {
    case 'PAYMENT_PENDING':
      return { percent: 15, label: 'Payment Pending', labelHi: 'भुगतान बाकी', colorClass: 'text-amber-600 bg-amber-50 dark:bg-amber-950/40 border-amber-300' };
    case 'PAYMENT_INITIATED':
      return { percent: 45, label: 'Payment Initiated', labelHi: 'भुगतान शुरू किया गया', colorClass: 'text-blue-600 bg-blue-50 dark:bg-blue-950/40 border-blue-300' };
    case 'PAYMENT_CONFIRMED':
      return { percent: 80, label: 'Payment Confirmed', labelHi: 'भुगतान सत्यापित', colorClass: 'text-indigo-600 bg-indigo-50 dark:bg-indigo-950/40 border-indigo-300' };
    case 'PAYMENT_COMPLETED':
      return { percent: 100, label: 'Payment Completed', labelHi: 'भुगतान संपन्न', colorClass: 'text-emerald-700 bg-emerald-50 dark:bg-emerald-950/40 border-emerald-300' };
    case 'PAYMENT_FAILED':
      return { percent: 40, label: 'Payment Failed', labelHi: 'भुगतान असफल', colorClass: 'text-rose-600 bg-rose-50 dark:bg-rose-950/40 border-rose-300' };
    case 'PAYMENT_CANCELLED':
      return { percent: 0, label: 'Payment Cancelled', labelHi: 'भुगतान रद्द', colorClass: 'text-slate-600 bg-slate-100 dark:bg-slate-800 border-slate-300' };
    case 'REFUND_PENDING':
      return { percent: 70, label: 'Refund Pending', labelHi: 'रिफंड प्रक्रिया में', colorClass: 'text-purple-600 bg-purple-50 dark:bg-purple-950/40 border-purple-300' };
    case 'REFUNDED':
      return { percent: 100, label: 'Directly Refunded', labelHi: 'सीधा रिफंड पूरा', colorClass: 'text-teal-700 bg-teal-50 dark:bg-teal-950/40 border-teal-300' };
    default:
      return { percent: 0, label: 'Unknown', labelHi: 'अज्ञात', colorClass: 'text-slate-500 bg-slate-50' };
  }
}
